### PR TITLE
Sync script for postfix address book (HPC-7565)

### PR DESCRIPTION
**Pull request by itkovian**

### CODE REVIEW

File: bin/sync_vsc_email_postfix.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        active_accounts, inactive_accounts = self.get_accounts()
+
+        active_emails = dict([(a.vsc_id, a.email) for a in active_accounts])
+
+        with open(self.options.postfix_address_book, 'r') as pab:
+            address_map = dict( [ (v, t) for [v, t] in [l.split() for l in pab.readlines()] ])
+
+        target_emails = address_map.update(active_emails)
+
+        with open(self.options.postfix_address_book, 'w') as pab:
```
Comment (itkovian) [inline bin/sync_vsc_email_postfix.py]: Should be written elsewhere and atomically updating the actual file.

Comment (stdweird) [reply bin/sync_vsc_email_postfix.py]: should we move the https://github.ugent.be/hpcugent/vsc-data/blob/master/bin/sync_vsc_data_gridmap.py#L86 to sync class?

File: bin/sync_vsc_email_postfix.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+            pab.write("\n".join(["%s    %s" % (k, v) for (k, v) in target_emails]))
+
+
+if __name__ == '__main__':
+    VscPostfixSync().main()
```
Comment (stdweird) [inline bin/sync_vsc_email_postfix.py]: add newline

File: bin/sync_vsc_email_postfix.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+        active_emails = dict([(a.vsc_id, a.email) for a in active_accounts])
+
+        with open(self.options.postfix_address_book, 'r') as pab:
+            address_map = dict( [ (v, t) for [v, t] in [l.split() for l in pab.readlines()] ])
+
+        target_emails = address_map.update(active_emails)
+
+        with open(self.options.postfix_address_book, 'w') as pab:
+            pab.write("\n".join(["%s    %s" % (k, v) for (k, v) in target_emails]))
```
Comment (stdweird) [inline bin/sync_vsc_email_postfix.py]: add trailing newline
`"%s %s" % x for x in target_emails` works too
what's up with all the spaces?

File: bin/sync_vsc_email_postfix.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+from vsc.accountpage.client import AccountpageClient
+from vsc.accountpage.sync import Sync
+from vsc.accountpage.wrappers import mkAccount
+
+
+class VscPostfixSync(Sync):
+    CLI_OPTIONS = {
+        postfix_address_book: ('Location of the postfix canonical address file', None, 'store', None),
+        token: ('OAuth token to access the account page', None, 'store', None)
```
Comment (stdweird) [inline bin/sync_vsc_email_postfix.py]: that is not needed, part of Sync class as (ususal) `access_token`

File: bin/sync_vsc_email_postfix.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+import sys
+
+from vsc.accountpage.client import AccountpageClient
+from vsc.accountpage.sync import Sync
+from vsc.accountpage.wrappers import mkAccount
+
+
+class VscPostfixSync(Sync):
+    CLI_OPTIONS = {
+        postfix_address_book: ('Location of the postfix canonical address file', None, 'store', None),
```
Comment (stdweird) [inline bin/sync_vsc_email_postfix.py]: `address_book` -> `alias_map`, and you can set a default `/etc/postfix/alias` or something like that

File: bin/sync_vsc_email_postfix.py (hunk 1 above)
Comment (itkovian) [reply bin/sync_vsc_email_postfix.py]: To vsc-utils at least :)

**Comment by itkovian** [comment]
Actually, we could just dump the complete _active_ list very time, unless that is considered too much traffic from the account page.

**Comment by stdweird** [comment]
@itkovian as already mentioned, look at the eppn code

File: bin/sync_vsc_email_postfix.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+from __future__ import print_function
+
+import logging
+
+from vsc.accountpage.sync import Sync
+
+
+class VscPostfixSync(Sync):
+    CLI_OPTIONS = {
+        'postfix_alias_map': ('Location of the postfix canonical map', None, 'store', '/etc/postfix/vsc_canonical'),
```
Comment (stdweird) [inline bin/sync_vsc_email_postfix.py]: change option name to canonical someting

File: bin/sync_vsc_email_postfix.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    def do(self, dryrun):
+        """Actual work.
+
+        - retrieve all the email addresss from changed users
+        - read the current file
+        - remove inactive users' email addresses
+        - add new/changed email addresses
+        - write file
+        """
+        active_accounts, inactive_accounts = self.get_accounts()
```
Comment (stdweird) [inline bin/sync_vsc_email_postfix.py]: if nothing is modifed on AP side, do nothing/ return empty list?

File: bin/sync_vsc_email_postfix.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+            address_set = set()
+    
+        address_set -= inactive_emails
+        address_map = dict(address_set)
+
+        address_map.update(active_emails)
+
+        with open(self.options.postfix_alias_map, 'w') as pab:
+            pab.write("\n".join(["%s@vscentrum.be %s" % kv for kv in address_map.items()]))
+            pab.write("\n")
```
Comment (stdweird) [inline bin/sync_vsc_email_postfix.py]: log that file was written

File: bin/sync_vsc_email_postfix.py (hunk 7 above)
Comment (stdweird) [reply bin/sync_vsc_email_postfix.py]: also log the changes?

File: bin/sync_vsc_email_postfix.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+        - retrieve all the email addresss from changed users
+        - read the current file
+        - remove inactive users' email addresses
+        - add new/changed email addresses
+        - write file
+        """
+        active_accounts, inactive_accounts = self.get_accounts()
+
+        active_emails = dict([(a.vsc_id, a.email) for a in active_accounts])
```
Comment (stdweird) [inline bin/sync_vsc_email_postfix.py]: why no set for both active and inactive mails, why not work with sets only?

File: bin/sync_vsc_email_postfix.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+        address_map = dict()
+
+        inactive_emails = set([(a.vsc_id, a.email) for a in inactive_accounts])
+        try:
+            with open(self.options.postfix_alias_map, 'r') as pab:
+                address_set = set([tuple(l.split()) for l in pab.readlines() if l])
+        except IOError as err:
+            logging.warning("No alias map at %s: %s", self.options.postfix_alias_map, err)
```
Comment (stdweird) [inline bin/sync_vsc_email_postfix.py]: canonical map

File: bin/sync_vsc_email_postfix.py [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+            address_set = set()
+    
+        address_set -= inactive_emails
+        address_map = dict(address_set)
+
+        address_map.update(active_emails)
+
+        with open(self.options.postfix_alias_map, 'w') as pab:
```
Comment (stdweird) [inline bin/sync_vsc_email_postfix.py]: out of curiosity, what is `pab` short of?

File: bin/sync_vsc_email_postfix.py [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+        """
+        active_accounts, inactive_accounts = self.get_accounts()
+
+        active_emails = dict([(a.vsc_id, a.email) for a in active_accounts])
+        address_map = dict()
+
+        inactive_emails = set([(a.vsc_id, a.email) for a in inactive_accounts])
+        try:
+            with open(self.options.postfix_alias_map, 'r') as pab:
+                address_set = set([tuple(l.split()) for l in pab.readlines() if l])
```
Comment (stdweird) [inline bin/sync_vsc_email_postfix.py]: strip the `@vscentum.be`? and add unittests

File: bin/sync_vsc_email_postfix.py (hunk 6 above)
Comment (itkovian) [reply bin/sync_vsc_email_postfix.py]: Aargh :D

File: bin/sync_vsc_email_postfix.py (hunk 9 above)
Comment (itkovian) [reply bin/sync_vsc_email_postfix.py]: How does one update a set of tuples where the key is the first tuple element in Python?

File: bin/sync_vsc_email_postfix.py (hunk 11 above)
Comment (itkovian) [reply bin/sync_vsc_email_postfix.py]: postfix alias something :D will change.

File: bin/sync_vsc_email_postfix.py (hunk 12 above)
Comment (itkovian) [reply bin/sync_vsc_email_postfix.py]: We cannot add tests for things in bin?

File: bin/sync_vsc_email_postfix.py (hunk 9 above)
Comment (stdweird) [reply bin/sync_vsc_email_postfix.py]: ah, yes. i was confused with the copy paste below

but then the `-=` for inactive users also does not work (imagine a user that modified email and became inactive).
so the code should use same objects for everything.

for inactive users, you only need a set of the vsc accounts, not their email.
address_set should abe a dict too, and you can already filter out the inatcive users in the readlines loop

File: bin/sync_vsc_email_postfix.py (hunk 12 above)
Comment (stdweird) [reply bin/sync_vsc_email_postfix.py]: ofcourse we can, the `__main__` takes care of that

Comment (stdweird) [reply bin/sync_vsc_email_postfix.py]: https://github.ugent.be/hpcugent/vsc-data/blob/master/test/sync_gridmap.py

Comment (itkovian) [reply bin/sync_vsc_email_postfix.py]: That is new to me, we used to be unable to import those, afaik.

File: bin/sync_vsc_email_postfix.py (hunk 9 above)
Comment (itkovian) [reply bin/sync_vsc_email_postfix.py]: Ah yes, I see your point. It's a small window, but possible indeed.

File: bin/sync_vsc_email_postfix.py (hunk 12 above)
Comment (stdweird) [reply bin/sync_vsc_email_postfix.py]: the vsc-install 00 tests import everything we ship, since forever

File: bin/sync_vsc_email_postfix.py [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+                address_set = set([tuple(l.split()) for l in cm.readlines() if l])
+        except IOError as err:
+            logging.warning("No canonical map at %s: %s", self.options.postfix_canonical_map, err)
+
+        address_set -= inactive_emails
+        address_map = dict(address_set)
+        address_map.update(active_emails)
+
+        txt = "\n".join(["%s %s" % kv for kv in address_map.items()])
+        txt += "\n"
```
Comment (stdweird) [inline bin/sync_vsc_email_postfix.py]: add a `+['']` in the previous join

File: bin/sync_vsc_email_postfix.py [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+        - write file
+        """
+        active_accounts, inactive_accounts = self.get_accounts()
+
+        if not active_accounts and not inactive_accounts:
+            logging.info("No changed accounts. Not rewriting the canonical map file.")
+            return
+
+        active_emails = dict([("%s@vscentrum.be" % a.vsc_id, a.email) for a in active_accounts])
+        inactive_emails = set([("%s@vscentrum.be" % a.vsc_id, a.email) for a in inactive_accounts])
```
Comment (stdweird) [inline bin/sync_vsc_email_postfix.py]: still same cornercase with `-=`, and not needed at all. make a list of vsc@vscentrum address, you don't need the values

File: bin/sync_vsc_email_postfix.py [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+
+        logging.debug("active emails: %s" % active_emails)
+        logging.debug("inactive emails: %s" % inactive_emails)
+
+        address_map = dict()
+        address_set = set()
+        try:
+            with open(self.options.postfix_canonical_map, 'r') as cm:
+                address_set = set([tuple(l.split()) for l in cm.readlines() if l])
```
Comment (stdweird) [inline bin/sync_vsc_email_postfix.py]: `address = dict([tuple(l) for l in map(split, cm.readlines()) if l[0] not in inactive_emails])`
and you have your dict

File: bin/sync_vsc_email_postfix.py [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+            logging.info("No changed accounts. Not rewriting the canonical map file.")
+            return
+
+        active_emails = dict([("%s@vscentrum.be" % a.vsc_id, a.email) for a in active_accounts])
+        inactive_emails = set(["%s@vscentrum.be" % a.vsc_id for a in inactive_accounts])
+
+        logging.debug("active emails: %s" % active_emails)
+        logging.debug("inactive emails: %s" % inactive_emails)
+
+        address_map = dict()
```
Comment (stdweird) [inline bin/sync_vsc_email_postfix.py]: not needed. it's odd there's no `unused` prospecotr error...

File: bin/sync_vsc_email_postfix.py [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+        address_cur = dict()
+        try:
+            with open(self.options.postfix_canonical_map, 'r') as cm:
+                address_cur = dict(
+                    [tuple(l) for l in map(str.split(), cm.readlines()) if l and l[0] not in inactive_emails]
+                )
+        except IOError as err:
+            logging.warning("No canonical map at %s: %s", self.options.postfix_canonical_map, err)
+
+        address_map = dict(address_cur)
```
Comment (stdweird) [inline bin/sync_vsc_email_postfix.py]: remove `dict` (or remove address_map entriely, just update address_cur?)

File: test/sync_vsc_email.py [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+
+        newmap = [
+            "vsc40001@vscentrum.be huppel@bar.com",
+            "vsc40002@vscentrum.be foo2@bar.com",
+            "",
+        ]
+
+        vps.do(False)
+        with open(cmfn, 'r') as c:
+            self.assertEqual(c.read(), "\n".join(newmap))
```
Comment (wdpypere) [inline test/sync_vsc_email.py]: newline

**Comment by wdpypere** [comment]
needs a version bump

**Comment by itkovian** [comment]
Done.

**Pull request merged**